### PR TITLE
Refactor filters to support iterator operation

### PR DESCRIPTION
**Pull request by Arachnid**
This PR introduces a `FindOnce` method for filters. `FindOnce` finds the next block that matches the filter and returns all matching logs from that block. If there are no further matching logs, it returns a nil slice. This method allows callers to iterate over large sets of logs progressively.

The changes for this PR introduce a small inefficiency relating to mipmaps: the first time a filter is called, it acts as if all mipmaps are matched, and thus iterates several blocks near the requested start point. This is in the interest of simplicity and avoiding duplicate mipmap lookups each time `FindOnce` is called.

### CODE REVIEW

File: eth/filters/filter.go [hunk 1]
```diff
@@ -85,8 +85,7 @@ func (f *Filter) SetTopics(topics [][]common.Hash) {
 	f.topics = topics
 }
 
-// Run filters logs with the current parameters set
-func (f *Filter) Find(ctx context.Context) ([]*vm.Log, error) {
+func (f *Filter) FindOnce(ctx context.Context) ([]*vm.Log, error) {
```
Comment (bas-vk) [inline eth/filters/filter.go]: Add some documentation.

Comment (Arachnid) [inline eth/filters/filter.go]: Done.

**Comment by Arachnid** [comment]
> Please run go fmt or something similar.

Done.

> Did some basic testing and it seems to work. I assume this PR enables users to use the filter system a bit more efficient in their own applications or on an "offline" chain?

This is a prelude to adding support for retrieving historical log entries from streaming filters.

> Would also be nice to add documentation on filters.New when the mip map filter can be used and when not.

Done.

File: eth/filters/filter.go [hunk 2]
```diff
@@ -54,6 +54,8 @@ type Filter struct {
 
 // New creates a new filter which uses a bloom filter on blocks to figure out whether
 // a particular block is interesting or not.
+// MipMaps allow past blocks to be searched much more efficiently, and can be enabled
+// any time a filter includes addresses to filter on.
 func New(backend Backend, useMipMap bool) *Filter {
```
Comment (bas-vk) [inline eth/filters/filter.go]: I would only mention that the mip map option can't be used with a light client and leave the address remark out. The light client doesn't create the required bloom filters. The code can figure out that his option is only useful for addresses, its not something the client needs to know.

Maybe its better to remove this option and always try to fetch the bloom filter. If that fails fallback to the less efficient method. But that is something for a separate PR.

**Pull request merged**